src/mlpack/methods/ann/loss_functions/cosine_embedding_loss.hpp

```diff
@@ -0,0 +1,147 @@
+/**
+ * @file cosine_embedding_loss.hpp
+ * @author Kartik Dutt
+ *
+ * Definition of the Cosine Embedding loss function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LOSS_FUNCTION_COSINE_EMBEDDING_HPP
+#define MLPACK_METHODS_ANN_LOSS_FUNCTION_COSINE_EMBEDDING_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Cosine Embeddings Loss function is used for measuring whether two inputs are
+ * similar or dissimilar, using the cosine distance, and is typically used
+ * for learning nonlinear embeddings or semi-supervised learning.
+ *
+ * @f{eqnarray*}{
+ * f(x) = 1 - cos(x1, x2) , for y = 1
+ * f(x) = max(0, cos(x1, x2) - margin) , for y = -1
+ * @f}
+ * 
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class CosineEmbeddingLoss
+{
+ public:
+  /**
+   * Create the CosineEmbeddingLoss object.
+   * 
+   * @param margin Increases cosine distance in case of dissimilarity.
+   *               Refer definition of cosine-embedding-loss above.
+   * @param takeMean Boolean variable to specify whether to take mean or not.
+   *                 Specifies reduction method i.e. sum or mean corresponding
+   *                 to 0 and 1 respectively. Default value = 0.
+   */
+  CosineEmbeddingLoss(const double margin = 0.0, const bool takeMean = false);
+
+  /**
+   * Ordinary feed forward pass of a neural network.
+   *
+   * @param x1 Input data used for evaluating the given function.
+   * @param x2 Input data used for evaluating the given function.
+   * @param y Input data used to determine whether to calculate
+   *          cosine similarity or dissimilarity. It should only
+   *          contain values equal to 1 or -1.
+   */
+  template<
+    typename FirstInputType,
+    typename SecondInputType,
+    typename ThirdInputType
+  >
+  double Forward(const FirstInputType& x1, const SecondInputType& x2,
+      const ThirdInputType& y);
```

Comment:
I'm a bit confused; this doesn't match the existing ANN layer API.  How will this work with the rest of the codebase?

Reply:
Hmm, Agreed. I think this can be used with a custom layer before it. Like we can have a vector of LayerTypes and use it for pairwise comparison between their outputs. However after your comment I think there is a way to improve it so a new layer isn't required. I can use y (scalar) as hyper parameter so that at one time the layer represents similarity / dissimilarity rather than supporting both. This is a bit different from pytorch but I think it makes sense do this. If a user wants to create a layer exactly like he / she can create a wrapper over this. I think I will make the change and then you could take a look. Thanks.

Reply:
Can you combine both inputs in a matrix?

Reply:
I think I can do that, then we could split them up in forward and backward and perform the operation so the user would have to use concatenate before this right?
I am assuming this is what you had in mind, and I guess it makes sense so. A similar thing could be done with margin ranking loss, triplet loss and so on. Looking forward to response. Thanks a lot.

Reply:
Hmm, the new strategy looks fine to me, but it differs from the PyTorch implementation in that it only allows a single value to be specified for `y` (now called `similarity`) instead of one for each point.  However I'm not clear on what exact application would need such support, and I can't seem to find an example that demonstrates that usage in a real-world situation.  So, I think the `similarity` approach that you've done here is nice and meshes nicely with the rest of the API, but maybe we are overlooking some use case.  I'm not sure.  (My comment is mostly just a thought, not a request. :))

Reply:
Agreed, I don't think point wise `similarity` has a particular use case but we might need to think of a way to add functions like Margin Ranking Loss which would require three inputs. What do you think?

Reply:
I'm not familiar with the margin ranking loss so I'm probably the wrong person to ask---I might suggest a "workaround" in terms of having the `Forward()` function take a concatenated vector (i.e. instead of giving `a` and `b`, give `[a, b]`).  However I'm not sure how nice that is, so maybe there's a better idea.

Reply:
Yes, I think that's what @zoq suggested. I guess that works for now. Thanks. I will mention it in the other PRs.